src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/CompilationTagHelperResolver.cs

```diff
@@ -59,20 +58,22 @@ internal class CompilationTagHelperResolver(ITelemetryReporter? telemetryReporte
         static void ExecuteProviders(ITagHelperDescriptorProvider[] providers, TagHelperDescriptorProviderContext context, ITelemetryReporter? telemetryReporter)
         {
             using var _ = StopwatchPool.GetPooledObject(out var watch);
-            using var timingDictionary = new PooledDictionaryBuilder<string, object?>();
+            using var properties = new PooledArrayBuilder<Property>(providers.Length);
 
             foreach (var provider in providers)
             {
                 watch.Restart();
                 provider.Execute(context);
                 watch.Stop();
 
-                var propertyName = $"{provider.GetType().Name}.elapsedtimems";
-                Debug.Assert(!timingDictionary.ContainsKey(propertyName));
-                timingDictionary.Add(propertyName, watch.ElapsedMilliseconds);
+                if (telemetryReporter is not null)
+                {
+                    var propertyName = $"{provider.GetType().Name}.elapsedtimems";
+                    properties.Add(new(propertyName, watch.ElapsedMilliseconds));
+                }
             }
 
-            telemetryReporter?.ReportEvent("taghelperresolver/gettaghelpers", Severity.Normal, timingDictionary.ToImmutable());
+            telemetryReporter?.ReportEvent("taghelperresolver/gettaghelpers", Severity.Normal, properties.DrainToImmutable());
```

Comment:
we can just report this in the telemetry reporter null check

Reply:
That will produce a telemetry event per provider, which is different than what we do now. Do we want this telemetry event to be fired 8-ish times rather than once?

Reply:
Oh, I missed that it was in a loop. Carry on